tremor-script/src/interpreter.rs

```diff
@@ -693,99 +741,138 @@ where
     'event: 'run,
 {
     let patch_expr = expr;
+    let mut evaluated: Vec<PreEvaluatedPatchOperation> = Vec::with_capacity(expr.operations.len());
+    // first pass over the operations, evaluating them
+    // and (IMPORTANT) get it into an owned, possibly cloned value, so we reference
+    // the target value in the state before any patch operation has been executed.
     for op in &expr.operations {
-        // NOTE: This if is inside the for loop to prevent obj to be updated
-        // between iterations and possibly lead to dangling pointers
+        evaluated.push(match op {
+            PatchOperation::Insert { ident, expr } => PreEvaluatedPatchOperation::Insert {
+                ident: stry!(ident.eval_to_string(opts, env, event, state, meta, local)),
+                ident_expr: ident,
+                value: stry!(expr.run(opts, env, event, state, meta, local)).into_owned(),
+            },
+            PatchOperation::Update { ident, expr } => PreEvaluatedPatchOperation::Update {
+                ident: stry!(ident.eval_to_string(opts, env, event, state, meta, local)),
+                ident_expr: ident,
+                value: stry!(expr.run(opts, env, event, state, meta, local)).into_owned(),
+            },
+            PatchOperation::Upsert { ident, expr } => PreEvaluatedPatchOperation::Upsert {
+                ident: stry!(ident.eval_to_string(opts, env, event, state, meta, local)),
+                value: stry!(expr.run(opts, env, event, state, meta, local)).into_owned(),
+            },
+            PatchOperation::Erase { ident } => PreEvaluatedPatchOperation::Erase {
+                ident: stry!(ident.eval_to_string(opts, env, event, state, meta, local)),
+            },
+            PatchOperation::Copy { from, to } => PreEvaluatedPatchOperation::Copy {
+                from: stry!(from.eval_to_string(opts, env, event, state, meta, local)),
+                to: stry!(to.eval_to_string(opts, env, event, state, meta, local)),
+            },
+            PatchOperation::Move { from, to } => PreEvaluatedPatchOperation::Move {
+                from: stry!(from.eval_to_string(opts, env, event, state, meta, local)),
+                to: stry!(to.eval_to_string(opts, env, event, state, meta, local)),
+            },
+            PatchOperation::Merge { ident, expr } => PreEvaluatedPatchOperation::Merge {
+                ident: stry!(ident.eval_to_string(opts, env, event, state, meta, local)),
+                ident_expr: ident,
+                merge_value: stry!(expr.run(opts, env, event, state, meta, local)).into_owned(),
+            },
+            PatchOperation::TupleMerge { expr } => PreEvaluatedPatchOperation::TupleMerge {
+                merge_value: stry!(expr.run(opts, env, event, state, meta, local)).into_owned(),
+            },
+        })
+    }
+
+    // second pass over pre-evaluated operations
+    // executing them against the actual target value
+    for const_op in evaluated.drain(..) {
```

Comment:
Why `drain(..)` instead of moving the `evaluated` into the `for` loop (implicitly turning it `into_iter`)?

Reply:
good catch, the vec is going to be dropped after this function anyways :)